lib/improver/spotdata/neighbour_finding.py

```diff
@@ -0,0 +1,445 @@
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Neighbour finding for the Improver site specific process chain."""
+
+import numpy as np
+import cartopy.crs as ccrs
+# from iris.analysis.trajectory import interpolate
+from improver.spotdata.ancillaries import data_from_ancillary
+from improver.spotdata.common_functions import (ConditionalListExtract,
+                                                nearest_n_neighbours,
+                                                get_nearest_coords,
+                                                index_of_minimum_difference,
+                                                list_entry_from_index,
+                                                node_edge_test)
+
+
+class PointSelection(object):
+    """
+    For the selection of source data from a grid for use in deriving
+    conditions at an arbitrary coordinate.
+
+    """
+
+    def __init__(self, method='default'):
+        """neighbour_list = find_nearest_neighbours(cube, spot_sites)
+        The class is called with the desired method to be used in determining
+        the grid point closest to sites of interest.
+
+        """
+        self.method = method
+
+    def process(self, cube, sites, **kwargs):
+        """
+        Call the correct function to enact the method of PointSelection
+        specified.
+
+        """
+        function = getattr(self, self.method)
+        return function(cube, sites, **kwargs)
+
+    @staticmethod
+    def fast_nearest_neighbour(cube, sites, ancillary_data=None):
+        '''
+        Use iris coord.nearest_neighbour_index function to locate the nearest
+        grid point to the given latitude/longitude pair.
+
+        Performed on a 2D-surface; consider using the much slower
+        iris.analysis.trajectory.interpolate method for a more correct nearest
+        neighbour search with projection onto a spherical surface; this is
+        typically much slower.
+
+        Args:
+        -----
+        cube           : Iris cube of gridded data.
+        sites          : Dictionary of site data, including lat/lon and
+                         altitude information.
+                         e.g. {<site_id>: {'latitude': 50, 'longitude': 0,
+                                           'altitude': 10}}
+        ancillary_data : A dictionary containing additional model data that
+                         is needed. e.g. {'orography': <cube of orography>}
+
+        Returns:
+        --------
+        neighbours: Numpy array of grid i,j coordinates that are nearest to
```

Comment:
This should include a description of what the fourth item, which sometimes you've referred to as `edgecase` represents.